Make new pseudo to allow Mail to turn off attachment controls.

megangardner opened the pull request:
#### 607499977734f77ef66a84a9df7f9b2bf4a03278
<pre>
Make new pseudo to allow Mail to turn off attachment controls.
<a href="https://bugs.webkit.org/show_bug.cgi?id=243229">https://bugs.webkit.org/show_bug.cgi?id=243229</a>

Reviewed by Wenson Hsieh and Devin Rousso.

Mail would like to have the option to turn off attachment controls for some images.
Since this is stored in the shadowdom, we need to add a pseudo to allow mail to change
the style of the controls, to make them invisible.

We also need to expand the prefixes for custom pseudo to keep the compartmentalization
of these pseudos correct.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::parsePseudoElementType):
* Source/WebCore/dom/mac/ImageControlsMac.cpp:
(WebCore::ImageControlsMac::createImageControls):
* Source/WebCore/html/shadow/ShadowPseudoIds.cpp:
(WebCore::ShadowPseudoIds::appleAttachmentControlsContainer):
* Source/WebCore/html/shadow/ShadowPseudoIds.h:

Canonical link: <a href="https://commits.webkit.org/252862@main">https://commits.webkit.org/252862@main</a>
</pre>

whsieh reviewed (approved):
We should add a test for this at some point.

dcrousso commented on Source/WebCore/dom/mac/ImageControlsMac.cpp:
```diff
@@ -105,6 +106,7 @@ void createImageControls(HTMLElement& element)
     auto button = HTMLButtonElement::create(HTMLNames::buttonTag, element.document(), nullptr);
     button->setIdAttribute(imageControlsButtonIdentifier());
     controlLayer->appendChild(button);
+    controlLayer->setPseudo(ShadowPseudoIds::appleAttachmentControlsContainer());
```
Why not name this something like `-apple-image-controls` or `-webkit-image-controls`?  This doesn't seem like it's specific to `<attachment>` (and the `-container` part is kinda an implementation detail, not something we necessarily need to expose to websites) 🤔

whsieh replied:
It's used for `attachment` and attachment-backed `img` elements, so I think the `-attachment-` part of the pseudo element name is appropriate.

megangardner replied:
The plan is to rename the image stuff to attachment anyways because it also covers PDF.

megangardner committed:
Make new pseudo to allow Mail to turn off attachment controls.
https://bugs.webkit.org/show_bug.cgi?id=243229

Reviewed by Wenson Hsieh and Devin Rousso.

Mail would like to have the option to turn off attachment controls for some images.
Since this is stored in the shadowdom, we need to add a pseudo to allow mail to change
the style of the controls, to make them invisible.

We also need to expand the prefixes for custom pseudo to keep the compartmentalization
of these pseudos correct.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::parsePseudoElementType):
* Source/WebCore/dom/mac/ImageControlsMac.cpp:
(WebCore::ImageControlsMac::createImageControls):
* Source/WebCore/html/shadow/ShadowPseudoIds.cpp:
(WebCore::ShadowPseudoIds::appleAttachmentControlsContainer):
* Source/WebCore/html/shadow/ShadowPseudoIds.h:

Canonical link: https://commits.webkit.org/252862@main
